### ISSUE TITLE
Support IterableDataset

### ISSUE THREAD

**Issue by HamedHemati** [ContinualAI/avalanche#1159]
Hi everyone

I'm opening this issue to discuss and share ideas about how to build infinite real-time streams in Avalanche.

In simple words, the idea is to add the ability to create real-time (or technically infinite) online CL streams in Avalanche. There are good reasons for having such streams:
- if we have a very large dataset that we can't read at once
- if we want to retrieve data and create CL streams from remote sources such as real-time data streams

In the figure below, I'm showing the high-level view of the components that need to interact to generate such streams. 
Assuming that we have a PyTorch [`IterableDataset`](https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset), we can create the stream on the fly through an infinite loop. To generate an `OCLExperience`, a "retriever" function iterates-over/waits-for the next `N` samples needed to create the experience. If `N=1`, the experience can be considered online, and if `N>0` is a small number, it is usually referred to as pseudo-online. Once the experience is ready, it will be used to run `train_exp` inside the `BaseTemplate`.

![inf_streams](https://user-images.githubusercontent.com/11629528/196978431-3b644e50-fd16-4ec6-8b1b-0d739c79a526.png)

Now, an important remaining question is: which parts need to change in the base template?

The only part that I guess needs to be changed is the `train()` function inside `BaseTemplate`.

Currently, the train function looks as below:

```python
  for self.experience in experiences:
      self._before_training_exp(**kwargs)
      self._train_exp(self.experience, eval_streams, **kwargs)
      self._after_training_exp(**kwargs)
```

This part can be changed to something like:

```python
  while True:
      experience = get_next_experience()
      if experience is None:
          break
      self._before_training_exp(**kwargs)
      self._train_exp(self.experience, eval_streams, **kwargs)
      self._after_training_exp(**kwargs)
```

For a regular list of experiences, the  `get_new_experience` just iterates over the list and returns `None` once all items are returned, and for  IterableDatasets, it will wait until the OCLExperience is constructed, then returns the experience; if there is a termination signal from the main source it will return `None,` and the experience loop will end.

For the eval stream, it can remain as it is, or it can also be constructed over time, but that should be just additional details.

What do you think about the general picture, and what kind of problems do you see with this approach?

**Comment by HamedHemati** [comment]
Ok, I see. So, for the first case, your suggestion is that, similar to the online stream generator, we can define a scenario generator whose `train_stream` is a lazy iterator over experiences created in real time using the data retrieved from an iterable dataset. I think, that makes more sense than pushing it to the `train()` function inside `BaseTemplate` considering the definition of "experience" in Avalanche.  

For the second case, as you said, it's currently not possible. I also checked the implementation of `AvalancheDataset` is a subclass of `FlatData`, and `FlatData` checks the dataset length for returning items.  

So, to summarize it, we can divide this issue into two parts

- Creating an Infinite stream from a single iterable dataset -> only needs a new stream generator, and everything remains the same.
- A stream of iterable datasets -> requires the support for IterableDataset in the definition of `AvalancheDataset`, maybe something like `AvalancheIterableDataset` :D .

I think the second objective is more related to continual learning for a sequence of super large datasets, and that's more a matter of scalability, so I would say for now we can target the first one.

**Comment by AntonioCarta** [comment]
> I think the second objective is more related to continual learning for a sequence of super large datasets, and that's more a matter of scalability, so I would say for now we can target the first one.

Exactly. We can do it as soon as there is a need for it, but right now I don't see it. All the datasets used in the CL literature are not `IterableDataset`s.

**Comment by HamedHemati** [comment]
Currently not a priority. I'm closing this issue.
